Filter Core Hook and Logic

jason-ransom-slalom opened the pull request:
This PR is for adding the base filter system logic. The code folders have been structured in a way to allow scaling the amount of available filters and add UI components. UI stories and demos will be added in the future. Also note that small libraries and test/formatting configs were added or updated. If approved, this should close #77 .

JonTheBurley commented on src/Filter/modules/AdditionalJSONFilter.tsx:
```diff
@@ -0,0 +1,29 @@
+ * This filter intentionally does NOT implement getFilterValueFromApiPostBody, because the useFilter hook
+ * sets this filter's value based on the JSON "leftover" after all the other filters parse their values from JSON
+ */
+export const AdditionalJSONFilter = (): FilterModule<any> => ({
```
The FilterModule<any> looks like it really wants it to be typed as something closer to FilterPostBody or a generic JSON interface. Maybe we can craft something better for it so anyone using it understands a bit more what kind of data this operates on.

jason-ransom-slalom replied:
Updated

JonTheBurley reviewed (approved):
Should it be exported with the rest of the components in the index.ts file at this point? There are a lot of pieces exported, does it makes sense to have some sort of exported object that contains all of the filter exports to be exported in one piece for use outside this project?

jason-ransom-slalom commented:
> Should it be exported with the rest of the components in the index.ts file at this point? There are a lot of pieces exported, does it makes sense to have some sort of exported object that contains all of the filter exports to be exported in one piece for use outside this project?

@JonTheBurley  - Good point. Yes it should be. I'll work on this and have you take a look before merging.

jason-ransom-slalom commented:
@JonTheBurley - I created a [placeholder file for the upcoming Filter component](https://github.com/ToyotaResearchInstitute/lakefront/blob/eb47ecb725aac729300e867e65b36ce494bbb999/src/Filter/Filter.tsx), added [src/Filter/index.ts](https://github.com/ToyotaResearchInstitute/lakefront/blob/eb47ecb725aac729300e867e65b36ce494bbb999/src/Filter/index.ts), and updated the [exports file](https://github.com/ToyotaResearchInstitute/lakefront/blob/eb47ecb725aac729300e867e65b36ce494bbb999/src/index.ts).

This would allow imports in the following ways (though I'm sure we'll want to drop the "dist" eventually:

`import { Filter, useFilter, FILTER_MODE_OPTIONS, FilterHooks, FilterSet } from '@toyota-research-institute/lakefront';`

`import Filter, { useFilter, FILTER_MODE_OPTIONS, FilterHooks, FilterSet } from '@toyota-research-institute/lakefront/dist/Filter';`

`import { AdditionalJSONFilter } from '@toyota-research-institute/lakefront/dist/Filter/modules';`

What do you think?

michaelpuzon commented on src/index.ts:
```diff
@@ -1,15 +1,33 @@
 export { default as Collapsible } from './Collapsible/Collapsible';
+
+export { default as Filter } from './Filter';
+export * from './Filter'
```
Is this export * from Filter intentional?

jason-ransom-slalom replied:
@michaelpuzon - Yes, though it just changed again slightly. After some discussion with @JonTheBurley, we think it makes sense to export the util functions, types, and filter modules from the root. We need a way to expose those, but don't see an immediate need to scope them further (which is what I was originally attempting to do). Let me know if that needs further clarification.

jason-ransom-slalom replied:
With the most recent change, imports will work as follows:
`import { Filter, useFilter, FILTER_MODE_OPTIONS, getApiPostBody, getDefaultValue } from '@toyota-research-institute/lakefront';`

JonTheBurley replied:
I think that looks good. If we decide we need them separated later, we can work on that. I don't think it matters too much to the user how it looks when it's imported at the moment.

michaelpuzon replied:
My argument against this was finding it harder to debug later on. 
Let's you had this statement `import { getDefaultValue } from '@toyota-research-institute/lakefront';`
It's not that obvious that this was in the Filter part of lakefront. 
The other point is that it might be easy to have duplicate hooks and functions exported. An example would be having a future `getDefaultValue` function in a future module in lakefront.
I don't have a better solution right now. If you can think of a better solution down the road, feel free to revisit this.
